File: frontend/src/app/core/services/api-endpoint.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ApiEndpointResponse {
  id: string;
  name: string;
  url: string;
  httpMethod: string;
  checkIntervalSeconds: number;
  expectedStatusCode: number;
  description: string | null;
  createdAt: string;
  lastStatus: string | null;
  lastResponseTimeMs: number | null;
  lastCheckedAt: string | null;
}

export interface CreateApiRequest {
  name: string;
  url: string;
  httpMethod: string;
  checkIntervalSeconds: number;
  expectedStatusCode: number;
  description?: string;
}

export interface UpdateApiRequest {
  name: string;
  url: string;
  httpMethod: string;
  checkIntervalSeconds: number;
  expectedStatusCode: number;
  description?: string;
}

@Injectable({ providedIn: 'root' })
export class ApiEndpointService {
  private apiUrl = `${environment.apiUrl}/endpoints`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<ApiEndpointResponse[]> {
    return this.http.get<ApiEndpointResponse[]>(this.apiUrl);
  }

  getById(id: string): Observable<ApiEndpointResponse> {
    return this.http.get<ApiEndpointResponse>(`${this.apiUrl}/${id}`);
  }

  create(request: CreateApiRequest): Observable<ApiEndpointResponse> {
    return this.http.post<ApiEndpointResponse>(this.apiUrl, request);
  }

  update(id: string, request: UpdateApiRequest): Observable<ApiEndpointResponse> {
    return this.http.put<ApiEndpointResponse>(`${this.apiUrl}/${id}`, request);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/PulseWatch.Infrastructure/Services/ApiEndpointService.cs
using Microsoft.EntityFrameworkCore;
using PulseWatch.Application.DTOs.ApiEndpoint;
using PulseWatch.Application.Interfaces;
using PulseWatch.Domain.Entities;
using PulseWatch.Infrastructure.Data;

namespace PulseWatch.Infrastructure.Services;

public class ApiEndpointService : IApiEndpointService
{
    private readonly PulseWatchDbContext _context;

    public ApiEndpointService(PulseWatchDbContext context)
    {
        _context = context;
    }

    public async Task<List<ApiEndpointResponse>> GetAllAsync(Guid userId)
    {
        var endpoints = await _context.ApiEndpoints
            .Where(a => a.UserId == userId)
            .OrderByDescending(a => a.CreatedAt)
            .ToListAsync();

        var responses = new List<ApiEndpointResponse>();
        foreach (var ep in endpoints)
        {
            var lastLog = await _context.MonitoringLogs
                .Where(m => m.ApiEndpointId == ep.Id)
                .OrderByDescending(m => m.CheckedAt)
                .FirstOrDefaultAsync();

            responses.Add(MapToResponse(ep, lastLog));
        }

        return responses;
    }

    public async Task<ApiEndpointResponse?> GetByIdAsync(Guid id, Guid userId)
    {
        var ep = await _context.ApiEndpoints
            .FirstOrDefaultAsync(a => a.Id == id && a.UserId == userId);

        if (ep == null) return null;

        var lastLog = await _context.MonitoringLogs
            .Where(m => m.ApiEndpointId == ep.Id)
            .OrderByDescending(m => m.CheckedAt)
            .FirstOrDefaultAsync();

        return MapToResponse(ep, lastLog);
    }

    public async Task<ApiEndpointResponse> CreateAsync(CreateApiRequest request, Guid userId)
    {
        var endpoint = new ApiEndpoint
        {
            Id = Guid.NewGuid(),
            Name = request.Name,
            Url = request.Url,
            HttpMethod = request.HttpMethod,
            CheckIntervalSeconds = request.CheckIntervalSeconds,
            ExpectedStatusCode = request.ExpectedStatusCode,
            Description = request.Description,
            UserId = userId,
            CreatedAt = DateTime.UtcNow
        };

        _context.ApiEndpoints.Add(endpoint);
        await _context.SaveChangesAsync();

        return MapToResponse(endpoint, null);
    }

    public async Task<ApiEndpointResponse?> UpdateAsync(Guid id, UpdateApiRequest request, Guid userId)
    {
        var ep = await _context.ApiEndpoints
            .FirstOrDefaultAsync(a => a.Id == id && a.UserId == userId);

        if (ep == null) return null;

        ep.Name = request.Name;
        ep.Url = request.Url;
        ep.HttpMethod = request.HttpMethod;
        ep.CheckIntervalSeconds = request.CheckIntervalSeconds;
        ep.ExpectedStatusCode = request.ExpectedStatusCode;
        ep.Description = request.Description;

        await _context.SaveChangesAsync();

        var lastLog = await _context.MonitoringLogs
            .Where(m => m.ApiEndpointId == ep.Id)
            .OrderByDescending(m => m.CheckedAt)
            .FirstOrDefaultAsync();

        return MapToResponse(ep, lastLog);
    }

    public async Task<bool> DeleteAsync(Guid id, Guid userId)
    {
        var ep = await _context.ApiEndpoints
            .FirstOrDefaultAsync(a => a.Id == id && a.UserId == userId);

        if (ep == null) return false;

        _context.ApiEndpoints.Remove(ep);
        await _context.SaveChangesAsync();
        return true;
    }

    private static ApiEndpointResponse MapToResponse(ApiEndpoint ep, MonitoringLog? lastLog)
    {
        return new ApiEndpointResponse
        {
            Id = ep.Id,
            Name = ep.Name,
            Url = ep.Url,
            HttpMethod = ep.HttpMethod,
            CheckIntervalSeconds = ep.CheckIntervalSeconds,
            ExpectedStatusCode = ep.ExpectedStatusCode,
            Description = ep.Description,
            CreatedAt = ep.CreatedAt,
            LastStatus = lastLog != null ? (lastLog.IsSuccess ? "Healthy" : "Unhealthy") : "Pending",
            LastResponseTimeMs = lastLog?.ResponseTimeMs,
            LastCheckedAt = lastLog?.CheckedAt
        };
    }
}
